CMakeLists.txt

```diff
@@ -76,48 +76,58 @@ include(cmake/modules/SetROOTVersion.cmake)
 
 message(STATUS "Building ROOT version ${ROOT_FULL_VERSION}")
 
-#---Try to download a file to check internet connection-----------------------------------------
-message(STATUS "Checking internet connectivity")
-file(DOWNLOAD https://root.cern/files/cmake_connectivity_test.txt ${CMAKE_CURRENT_BINARY_DIR}/cmake_connectivity_test.txt
-  TIMEOUT 10 STATUS DOWNLOAD_STATUS
-)
-# Get the status code from the download status
-list(GET DOWNLOAD_STATUS 0 STATUS_CODE)
-# Check if download was successful.
-if(${STATUS_CODE} EQUAL 0)
-  # Succcess
-  message(STATUS "Checking internet connectivity - found")
-  # Now let's delete the file
-  file(REMOVE ${CMAKE_CURRENT_BINARY_DIR}/cmake_connectivity_test.txt)
-  set(NO_CONNECTION FALSE)
-else()
-  # Error
-  message(STATUS "Checking internet connectivity - failed: will not automatically download external dependencies")
-  set(NO_CONNECTION TRUE)
-endif()
-
-if(NO_CONNECTION)
-  if(clad AND fail-on-missing)
-    message(FATAL_ERROR "No internet connection. Please check your connection, or either disable the 'clad' option or the 'fail-on-missing' to automatically disable options requiring internet access")
-  elseif(clad)
-    message(STATUS "No internet connection, disabling the 'clad' option")
-  endif()
-  set(clad OFF CACHE BOOL "Disabled because there is no internet connection" FORCE)
-endif()
-
 #---Where to look first for cmake modules, before ${CMAKE_ROOT}/Modules/ is checked-------------
 set(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake/modules)
 
 #---Enable Folders in IDE like Visual Studio----------------------------------------------------
 set_property(GLOBAL PROPERTY USE_FOLDERS ON)
 
-#---Load some basic macros which are needed later for the confiuration and build----------------
+#---Load some basic macros which are needed later for the configuration and build----------------
 include(CheckCompiler)
 include(RootBuildOptions)
 include(RootMacros)
 include(CheckAssembler)
 include(CheckIntrinsics)
 
+#---Try to download a file to check internet connection-----------------------------------------
+
+# The connection check only makes sense for fail-on-missing=OFF, where the
+# result is used to decide whether to download a missing dependency as a builtin
+# from the internet, or to disable the feature that has the missing dependency.
+# With fail-on-missing=ON, it doesn't matter because disabling features is not
+# allowed. Therefore, we can skip the connection check to save some
+# configuration overhead and just assume we have internet: if a builtin can't be
+# downloaded there will be a configuration failure either way.
```

Comment:
What does the error look like in this case?  (Part of the connection test was too make the error 'readable/clear').

Reply:
I haven't tried this out, but the error will be some sort of download error when it attempts to download the builtin. Indeed that's a bit less clear, so the question here is really: do nicer error messages justify the connectivity check in the `fail-on-missing=ON` case? I would argue "no", because `fail-on-missing` is only used by experts anyway that will be able to also understand the download errors and make a decision whether they prefer to get internet or disable the feature that required this download. Maybe to help them out a little bit, we should mention which feature use a given builtin in the options documentation string for the `builtin_` options?

Reply:
> I haven't tried this out, 

We should know rather than guess.

> Do nicer error messages justify the connectivity check in the fail-on-missing=ON case? 

We might have a way (with more typing) to have both by adding the test (with 'if already check don't check again) in each of the code portion that will trigger a download (actually it might 'as simple as' adding a wrapper around `ExternalProject_Add`)

Reply:
Ah yes, that's a good idea! To "lazily" do the connection check only when a download is attempted :+1: I'll see how we can do this.

Reply:
Wait, it's not so easy actually, because some people pre-populate the sources that would be retrieved by `FetchContent` steered with CMake variable. So the logic to determine if the connection check needs to be done has to consider that too (see the discussion at the end of #11603). So I come back to my original conclusion: we should consider sacrificing the nicer error messages for a less fragile and simpler configuration.

In fact, right now it's already a problem because the connection check is always done and can trigger the `FATAL_ERROR` code patch in `SearchInstalledSoftware`, even if the `FetchContent` is sideloaded.

Reply:
> Wait, it's not so easy actually, because some people pre-populate the sources that would be retrieved by FetchContent

If we put it in the wrap we should be able to check if the files/directories already exist.

Reply:
I don't think this goes in the direction of making things less fragile and complicated, which was one of the considerations for the issue:
https://github.com/root-project/root/issues/11603#issuecomment-2102860108

In that case, we might as well implement the flag for the connectivity check:
https://github.com/root-project/root/issues/11603#issuecomment-1288852262